### PR TITLE
[NativeAOT] Make casting logic closer to CoreCLR

### LINKED ISSUE #84464: [NativeAOT] Possible perf improvements in casting
CoreClr appears to have faster casting helpers than NativeAOT. https://github.com/dotnet/runtime/pull/84430#issuecomment-1499671856

That means the NativeAOT helpers could likely be improved, since conceptually they should be applying similar logic.
It is worth compare the code of the helpers and related typesystem bits to see what can be copied/unified from whichever is a faster implementation.

### PR DESCRIPTION
**Pull request by VSadov**
Fixes: https://github.com/dotnet/runtime/issues/84464

The actual changes are not as big here as might seem. For the most part this is a refactoring of existing code to have a shape closer to CoreCLR cast helpers, so that similar patterns could be used - in a few places where that has not been done already in earlier changes.  
For example cases like `CheckCastAny` - could start with a cache lookup, since uncached code path can be complex and thus relatively slow. (also addresses some old TODOs in this area)

### CODE REVIEW

File: src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/TypeCast.cs [hunk 1]
```diff
@@ -37,21 +37,145 @@ internal static class TypeCast
         [Flags]
         internal enum AssignmentVariation
         {
-            Normal = 0,
+            /// <summary>
+            /// Conversion from an object. In the terminology of ECMA335 the relationship is called "compatible-with".
+            /// This is the relation used by castclass and isinst (III.4.3).
+            /// Value types are compatible with Object, ValueType and Enum (if applicable) and implemented interfaces.
+            /// </summary>
+            BoxedSource = 0,
```
Comment (VSadov) [inline src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/TypeCast.cs:45]: I have changed the enum here. Casting from objects is a much more common operation than reasoning about convertibility of types. When converting from objects the source is always in an object/boxed form, so it would be nice to have `variation == 0` for that case.

**Comment by jkotas** [comment]
Could you please collect numbers for casting microbenchmarks before/after this change?

File: src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs [hunk 2]
```diff
@@ -694,32 +694,29 @@ private ISymbolNode GetHelperFtnUncached(CorInfoHelpFunc ftnNum)
                     break;
 
                 case CorInfoHelpFunc.CORINFO_HELP_CHKCASTANY:
+                case CorInfoHelpFunc.CORINFO_HELP_CHKCASTARRAY:
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs:697]: You can delete `CORINFO_HELP_CHKCASTARRAY` and `CORINFO_HELP_ISINSTANCEOFARRAY` from JIT/EE interface. They are unnecessary now.

**Comment by VSadov** [comment]
One possible question here would be - Why not just port the CoreCLR casting helpers in their entirety and implement the internal calls in managed code - to have nearly the same implementation?  

I think one of the factor for the design of CoreClr casting logic is to avoid complicated API with the native type system. As a result that API is basically 2 internal calls methods - `IsInstanceOfAny_NoCacheLookup` and `ChkCastAny_NoCacheLookup`. 

On NativeAOT side various type system APIs are easily accessible, so we do not need to minimize the use of those APIs. On the other hand there are some differences, like the way we fetch the base type for arrays, that may stand in the way of code sharing.  
Thus I did not consider it is as a goal to share the code or making it maximally similar when possible, but I think I can be done.

**Comment by VSadov** [comment]
Also in NativeAOT these casting helpers have more users with additional needs, while in CoreClr it really just object casting.

**Comment by VSadov** [comment]
> Could you please collect numbers for casting microbenchmarks before/after this change?

I was thinking of what we could measure here. 
The most common cases of casting like regular object/interfaces casts did not change, so unlikely to see any differences. We may see differences in more complex cases - like casting to variant interfaces or arrays.

Perhaps just running just the regular casting benchmarks that perf lab uses would be informative enough about what changed perf-wise.

**Comment by VSadov** [comment]
I have run the same benchmark as in https://github.com/dotnet/runtime/pull/84430#issuecomment-1499671856

```cs
 internal class Program
    {
        const int iters = 1000000;

        static void Main(string[] args)
        {
            for(; ; )
            {
                Time(TestLStringToIROCstring);
            }
        }
        
        static void Time(Action a)
        {
            var sw = Stopwatch.StartNew();
            for (int i = 0; i < 100; i++)
            {
                a();
            }

            sw.Stop();
            System.Console.WriteLine(sw.ElapsedMilliseconds);
        }

        static object o = new List<string>();

        static void TestLStringToIROCstring()
        {
            for (int i = 0; i < iters; i++)
            {
                if (o as IReadOnlyCollection<object> == null)
                    throw null;

                if (o as IReadOnlyCollection<string> == null)
                    throw null;

                if (o as IEnumerable<object> == null)
                    throw null;

                if (o as IEnumerable<string> == null)
                    throw null;
            }
        }
    }
```

=== before the change
```
1451
1460
1456
1458
1470
1463
```

=== after the change:
```
799
801
798
799
800
799
```

The reason for the difference is that original code makes a number of calls. Profiler shows:
`MethodTable_get_IsArray`
`TypeCast__IsInstanceOf`
`TypeCast__IsInstanceOfVariantType`    -  this one does the cache lookup 

making calls and additional checks adds up.

In the new implementation there is only one helper call in the profile:
`TypeCast__IsInstanceOfAny`                - this one does the cache lookup

**Comment by VSadov** [comment]
For comparison the CoreCLR is a bit faster.

The same benchmark as above produces (smaller is better):
```
558
561
562
557
561
```

As I see in the debugger the native code that we run after this change is nearly the same between CoreCLR and NativeAOT. 
There are minor differences like loading of type pointers/handles. 
In JIT code they look like:
```asm
mov         rcx,7FFCA19E79E8h
```

In NativeAOT loading a type looks like:
```asm
lea         rcx, [rip + 0x73d65]
```

I do not see any other significant differences. Maybe it is just these little diffs and some indirect impact on code size or alignment that makes the difference.

**Comment by VSadov** [comment]
Another thing to notice as that compared to https://github.com/dotnet/runtime/pull/84430#issuecomment-1499671856 and prior to this change it looks like the benchmark in the comment has regressed.  

That was possibly caused by https://github.com/dotnet/runtime/pull/86029 . I suspect it made common cases faster, which is good, but regressed complex cases that rely on caching as more checks like introduction of `MethodTable_get_IsArray` could run before eventually hitting the cache. Just a guess though.

Anyways, it looks like after this PR the complex/cached case is faster than in https://github.com/dotnet/runtime/pull/84430#issuecomment-1499671856

File: src/coreclr/inc/corinfo.h [hunk 3]
```diff
@@ -429,12 +429,10 @@ enum CorInfoHelpFunc
     // the right helper to use
```
Comment (jkotas) [inline src/coreclr/inc/corinfo.h:429]: This also needs a new JIT/EE interface GUID.

cc @dotnet/jit-contrib

File: src/coreclr/jit/importer.cpp [hunk 4]
```diff
@@ -5469,18 +5469,18 @@ GenTree* Compiler::impCastClassOrIsInstToTree(
...
                 const bool isAbstract = (info.compCompHnd->getClassAttribs(pResolvedToken->hClass) &
-                                         (CORINFO_FLG_INTERFACE | CORINFO_FLG_ABSTRACT)) != 0;
+                    (CORINFO_FLG_INTERFACE | CORINFO_FLG_ABSTRACT)) != 0;
                 canExpandInline = !isAbstract;
             }
         }
-        else if ((helper == CORINFO_HELP_ISINSTANCEOFCLASS) || (helper == CORINFO_HELP_ISINSTANCEOFARRAY))
+        else if ((helper == CORINFO_HELP_ISINSTANCEOFCLASS) || (helper == CORINFO_HELP_CHKCASTANY))
         {
             // If the class is exact, the jit can expand the IsInst check inline.
```
Comment (jkotas) [inline src/coreclr/jit/importer.cpp:5485]: I think this should do isAbstract check like the above.

Comment (VSadov) [reply src/coreclr/jit/importer.cpp:5485]: As I understand `isClassExact` implies that the type is a nonvariant final. The interfaces that may make here will be either variant or use `CORINFO_HELP_CHKCASTINTERFACE`.

Comment (VSadov) [reply src/coreclr/jit/importer.cpp:5485]: That is - unless we want to do inline expansion for throwing and nonthrowing path under the same conditions. I kind of assumed the difference is intentional.

Comment (jkotas) [reply src/coreclr/jit/importer.cpp:5485]: I think the idea is that we want to inline the exact match if it has a chance of succeeding.

@EgorBo Do you have any additional insights about this logic?

Comment (VSadov) [reply src/coreclr/jit/importer.cpp:5485]: Also it looks like inline expansion of `isinst` generally does not call the helper, just produces `null` when types do not match. 
https://github.com/dotnet/runtime/blob/fff52666cf669946254b07764e464f75f0b69c38/src/coreclr/jit/importer.cpp#L5637

If that is the case, then "inlining" in abstract case is still the right thing to do - no need to call the helper. The intersection of `isClassExact` and `isAbstract` could probably just optimize the whole deal into `null`. Although it might be too rare to worry about.

I am also wondering how all this works with COM, type equivalency, ICastable and other similar things. How can we avoid calling the helper if the object to which `isinst` is applied could be one of those cases? 
The more I look at this the more confusing it gets...

Comment (jkotas) [reply src/coreclr/jit/importer.cpp:5485]: > I am also wondering how all this works with COM, type equivalency, ICastable and other similar thing

These things are handled by the slow paths of the helpers. It means that the ininst negative case is optimizable only when we can be sure that none of these rare cases are possible.

**Comment by VSadov** [comment]
There are some `superpmi` failures. I can't tell if that simply tells that there are codegen diffs or something is crashing.

**Comment by jakobbotsch** [comment]
> There are some `superpmi` failures. I can't tell if that simply tells that there are codegen diffs or something is crashing.

The SPMI failures are happening because of changes in the JIT-EE interface without a corresponding JIT-EE GUID update. Once you update the JIT-EE GUID SPMI will be quiet.

File: src/coreclr/jit/importer.cpp (hunk 4 above)
Comment (EgorBo) [reply src/coreclr/jit/importer.cpp:5485]: > I think this should do isAbstract check like the above.

My understanding that we try to avoid expansion for `foo is T` if `T` is not exact, but we currently do that for `(T)foo` because the latter, presumably, has less options and InvalidCastException is not expected to be popular.

We can, probably, enable it for the former as well if we collect some numbers on how often this check is taken, right?
For JIT, I think, we'll eventually enable PGO for casts by default.

Comment (EgorBo) [reply src/coreclr/jit/importer.cpp:5485]: E.g.:
```cs
public class ClassA { }
public class ClassB : ClassA { }

class Program
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    bool IsClassA(object obj) => obj is ClassA;

    static void Main()
    {
        Program p = new();
        for (int i = 0; i < 100; i++)
        {
            p.IsClassA(new ClassB());
            Thread.Sleep(16);
        }
    }
}
```
Current codegen for `IsClassA`:
```asm
; Assembly listing for method Program:IsClassA(System.Object):bool:this (Tier1)
       sub      rsp, 40
       mov      rcx, 0x7FFC89533648      ; ClassA
       call     [CORINFO_HELP_ISINSTANCEOFCLASS]
       test     rax, rax
       setne    al
       movzx    rax, al
       add      rsp, 40
       ret 
```
(and if we partially expand it for `ClassA` it will be a pure overhead)
Now, codegen with `DOTNET_JitProfileCasts=1`:
```asm
; Assembly listing for method Program:IsClassA(System.Object):bool:this (Tier1)
       sub      rsp, 40
       mov      rax, rdx
       test     rax, rax
       je       SHORT G_M61590_IG05
       mov      rcx, 0x7FFC89503328      ; ClassB
       cmp      qword ptr [rax], rcx
       je       SHORT G_M61590_IG05
       mov      rcx, 0x7FFC895031E0
       call     [CORINFO_HELP_ISINSTANCEOFCLASS]
G_M61590_IG05:
       test     rax, rax
       setne    al
       movzx    rax, al
       add      rsp, 40
       ret
```

**Comment by EgorBo** [comment]
@MihuBot

File: src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/RuntimeExports.cs [hunk 5]
```diff
@@ -392,26 +392,22 @@ internal static unsafe IntPtr RhGetRuntimeHelperForType(MethodTable* pEEType, Ru
                         return (IntPtr)(delegate*<MethodTable*, object>)&InternalCalls.RhpNewFast;
 
                 case RuntimeHelperKind.IsInst:
-                    if (pEEType->IsArray)
-                        return (IntPtr)(delegate*<MethodTable*, object, object>)&TypeCast.IsInstanceOfArray;
-                    else if (pEEType->HasGenericVariance)
-                        return (IntPtr)(delegate*<MethodTable*, object, object>)&TypeCast.IsInstanceOf;
+                    if (pEEType->IsArray || pEEType->HasGenericVariance)
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/RuntimeExports.cs:395]: We should not need the `IsArray` check here. It is covered by the `IsParameterizedType` check below.

This can be even possibly changed to `if (pEEType->HasGenericVariance || pEEType->IsParameterizedType || pEEType->IsFunctionPointerType)` so that there is only a one path that returns the Any helper.

**Review by jkotas — APPROVED** [review]
LGTM. Thank you!

File: src/coreclr/jit/importer.cpp (hunk 4 above)
Comment (VSadov) [reply src/coreclr/jit/importer.cpp:5485]: I am still confused at what makes the `IsInst` optimization correct for types that have special casting semantics - like `ICastable`, COM or something with type equivalence. 

`impIsClassExact` does not check for these cases. It only checks for "final" + "nonvariant" case (in array case, the element is checked)

If a "special" casting is always to an interface, then, yes, we will never optimize "special" scenarios as they would either use interface-specific helpers (`CHKCASTINTERFACE`/`ISINSTANCEOFINTERFACE`) or will be filtered out due to variance. 
Am not sure it is always to an interface though.  

Basically if we have 
```cs
bool converts = obj is SomeSealedClass;
```

and `obj` happens to convert to `SomeSealedClass` via ICastable or type equivalence, would we optimize this case incorrectly into a type comparison?

Comment (jkotas) [reply src/coreclr/jit/importer.cpp:5485]: Are you able to construct a repro that shows that this optimization is incorrect for these corner cases?

I think it is likely that you are right that this optimization is incorrect for some corner cases. The `impIsClassExact ` pattern where the JIT tries to figure some kind of non-trivial type system property by combining type flags on its own is bug prone. It is better to have these non-trivial type system properties to be computed by dedicated JIT/EE interface APIs.

Comment (jkotas) [reply src/coreclr/jit/importer.cpp:5485]: Also, it seems that some of these buggy corner cases could have been introduced by deleting `CORINFO_HELP_ISINSTANCEOFARRAY`/`CORINFO_HELP_CHKCASTARRAY` helpers. We should get closure on this before merging this PR.

Comment (VSadov) [reply src/coreclr/jit/importer.cpp:5485]: Yes, that is my worry - we may have subtle holes here and I may be making things worse. Even though tests are passing.

It is possible to check that the destination type is an array and match the behavior as if array-specific helpers are still there, but need to be more sure that the current behavior is correct.  

Checked casts are ok, since we call helpers on nonmatch, but isinst is a concern.

Comment (VSadov) [reply src/coreclr/jit/importer.cpp:5485]: I looked at the cases when an object may cast to/from something that is not in its inheritance. We have 3 cases:
- `ICastable`   - not a concern, since the destination type must be an interface (so it will use different helpers)
- `IDynamicCastable` - same as above, not a concern.
- Type Equivalence  - it seems to be possible, at least in theory, to cast to an otherwise unrelated and sealed type via type equivalence. 

**I think types that participate in type equivalence should be rejected by `impIsClassExact` per the helper's intent.** 
I've added a commit that implements that - https://github.com/dotnet/runtime/pull/89548/commits/3a38456b92864e3af45727a4c285d7a3fed44231

I, however, could not figure how to construct an example that would demonstrate a problem. While I see how type equivalence works in the type system, I do not completely understand how such type are created by programs.

I have tried the following, but it did not result in types compatible via equivalence. Perhaps something is missing?

```cs
using System.Runtime.InteropServices;

[assembly:ImportedFromTypeLib("qq")]

namespace ConsoleApp33
{
    [TypeIdentifier("a", "b")]
    struct T1
    {
    }

    [TypeIdentifier("a", "b")]
    struct T2
    {
    }

    internal class Program
    {
        static void Main(string[] args)
        {
            object o = new T1();
            Console.WriteLine(o is T2);
            Console.WriteLine(typeof(T1).IsEquivalentTo(typeof(T2)));

            T2 c = (T2) o;
            Console.WriteLine (c.ToString ());
        }
    }
}
```

Comment (jkotas) [reply src/coreclr/jit/importer.cpp:5485]: > I have tried the following, but it did not result in types compatible via equivalence. Perhaps something is missing?

The type names need to match. You need a helper assembly with the equivalent type to make this work. For example, look for `EquivalentValueType` under `src\libraries\System.Reflection\tests`.

Comment (VSadov) [reply src/coreclr/jit/importer.cpp:5485]: It appears to be possible to add a scenario to EquivalentValueType test that is sensitive to the `CORINFO_FLG_TYPE_EQUIVALENCE` fix.  

Another important part is to force `AggressiveOptimization` to make sure that JIT optimizes the cast.

File: src/libraries/System.Reflection/tests/GetTypeTests.cs [hunk 6]
```diff
@@ -296,6 +297,20 @@ public void TestTypeIdentifierAttribute()
             // Ensure we can invoke with an arg type that is duplicated in another assembly but having [TypeIdentifier].
             args = new object[1] { Activator.CreateInstance(otherEquivalentValueType) };
             Assert.Equal(42, mi.Invoke(null, args));
+
+            // ensure that an instance of otherEquivalentValueType can cast to EquivalentValueType
```
Comment (jkotas) [inline src/libraries/System.Reflection/tests/GetTypeTests.cs:301]: This is codegen test. It would be more appropriate to add it under src\tests\baseservices\typeequivalence. For example, you can add it to `src\tests\baseservices\typeequivalence\simple\Simple.cs`.

File: src/libraries/System.Reflection/tests/GetTypeTests.cs [hunk 7]
```diff
@@ -296,6 +297,20 @@ public void TestTypeIdentifierAttribute()
...
+            object otherEquivalentValueTypeInstance = Activator.CreateInstance(otherEquivalentValueType);
+            Assert.True(otherEquivalentValueTypeInstance is EquivalentValueType);
+            EquivalentValueType inst = (EquivalentValueType)otherEquivalentValueTypeInstance;
+
+            OptimizedMethod(otherEquivalentValueTypeInstance);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
```
Comment (jkotas) [inline src/libraries/System.Reflection/tests/GetTypeTests.cs:309]: This should not be needed. We have runs with tiering disabled, JIT stress, etc. that will take care of it.

Comment (VSadov) [reply src/libraries/System.Reflection/tests/GetTypeTests.cs:309]: The attribute will be useful to prevent cases when this scenario passes locally and fails in CI.
When I run the test locally it did not fail. I think it can lead to unnecessary confusion if it relies on testing environment.

Comment (jkotas) [reply src/libraries/System.Reflection/tests/GetTypeTests.cs:309]: The same problem exists for many thousands of existing codegen optimizations tests that we have in the repo.

**Comment by VSadov** [comment]
I'd be ok with merging this now. I think all concerns have been resolved. Let me know if there is something that may be missing.

File: src/coreclr/vm/jitinterface.cpp [hunk 8]
```diff
@@ -6058,8 +6061,7 @@ CorInfoHelpFunc CEEInfo::getCastingHelperStatic(TypeHandle clsHnd, bool fThrowin
             *pfClassMustBeRestored = true;
```
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:6061]: The `pfClassMustBeRestored` argument is dead code and it can be deleted. It only controls whether we end up calling `classMustBeLoadedBeforeCodeIsRun` that is a no-op.

File: src/coreclr/vm/jitinterface.cpp [hunk 9]
```diff
@@ -6058,8 +6061,7 @@ CorInfoHelpFunc CEEInfo::getCastingHelperStatic(TypeHandle clsHnd, bool fThrowin
...
         }
 
-        // If it is an array, use the fast array helper
-        helper = CORINFO_HELP_ISINSTANCEOFARRAY;
+        _ASSERTE(helper == CORINFO_HELP_ISINSTANCEOFANY);
     }
     else
     if (!clsHnd.IsTypeDesc() && !Nullable::IsNullableType(clsHnd))
```
Comment (jkotas) [inline src/coreclr/vm/jitinterface.cpp:6067]: The native AOT implementation does not checks for Nullable. Is this check redundant here or is the check for Nullable missing in native AOT?

File: src/coreclr/jit/importer.cpp [hunk 10]
```diff
@@ -13612,7 +13612,7 @@ methodPointerInfo* Compiler::impAllocateMethodPointerInfo(const CORINFO_RESOLVED
 bool Compiler::impIsClassExact(CORINFO_CLASS_HANDLE classHnd)
 {
     DWORD flags     = info.compCompHnd->getClassAttribs(classHnd);
-    DWORD flagsMask = CORINFO_FLG_FINAL | CORINFO_FLG_VARIANCE | CORINFO_FLG_ARRAY;
+    DWORD flagsMask = CORINFO_FLG_FINAL | CORINFO_FLG_VARIANCE | CORINFO_FLG_TYPE_EQUIVALENCE | CORINFO_FLG_ARRAY;
```
Comment (jkotas) [inline src/coreclr/jit/importer.cpp:13615]: Both `CORINFO_FLG_VARIANCE` and `CORINFO_FLG_TYPE_EQUIVALENCE` are only computed to make the `impIsClassExact` work. Computing these flags is a waste in all other cases. It is the kind of pattern that calls for introduction of dedicated JIT/EE interface API that replaces the flags.

File: src/coreclr/vm/jitinterface.cpp (hunk 9 above)
Comment (VSadov) [reply src/coreclr/vm/jitinterface.cpp:6051]: The difference is that JIT tries to transform simple cases like `o is int?` into `o is int`, but can't do that when needs a type lookup. As I understand that is due to limitations of IR.

https://github.com/dotnet/runtime/blob/40b39ff7df1dc2388a5865c9ad151dce88fa007d/src/coreclr/jit/importer.cpp#L5387-L5391

NativeAOT does not seem to have a problem to express such lookup, so we always cast with nullable stripped

https://github.com/dotnet/runtime/blob/17d13fa10be9020c851e8f2e668685b4cb98d6eb/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Compilation.cs#L301-L306

Comment (VSadov) [reply src/coreclr/vm/jitinterface.cpp:6051]: In other words in NativeAOT the type check will always be to an unboxed underlying type, thus CLASS variant is correct, and also optimizable, since T? is `isClassExact`.

In CoreCLR the type check in complex cases could be to the actual `Nullable<SomeStruct<string>>`, thus it should use ANY variant. It also means that we may be introducing an optimization bug, since such IsInst cannot be lowered to a type/handle compare.

I will check if that is the case.

Comment (VSadov) [reply src/coreclr/vm/jitinterface.cpp:6051]: Sadly that is the case. The following works incorrectly

```cs
===== Prints:

True
False


using System.Runtime.CompilerServices;

namespace ConsoleApp34
{
    struct S1<T>
    {
        public T value;
    }

    internal class Program
    {
        static object o;
        static void Main(string[] args)
        {
            Test<int>();
            Test<string>();
        }

        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
        private static void Test<T>()
        {
            o = new S1<T>();

            Console.WriteLine(o is S1<T>?);
        }
    }
}
```

This PR indirectly enabled casting optimizations for cases like `o is int?`, but we need to suppress it, since in more general cases it does not work correctly.

Comment (VSadov) [reply src/coreclr/vm/jitinterface.cpp:6051]: Unless there are better ideas, I am thinking of constraining isinst optimization for `CORINFO_HELP_ISINSTANCEOFANY` only if converting to an array type. So we do not keep finding more broken cases.

That may be too conservative, but we should probably stay closer to the preexisting behavior for now and consider if more cases can work in 9.0

Comment (jkotas) [reply src/coreclr/vm/jitinterface.cpp:6051]: I am thinking that this PR as a whole is too risky for 8.0. Are there parts that we think are critical to get into .NET 8?

Comment (VSadov) [reply src/coreclr/vm/jitinterface.cpp:6051]: I think the actual casting helper change was low risk as that is mostly refactoring to get more cases to hit the cache earlier. Touching the JIT appears to be a lot more fragile.

There is nothing really "critical" for 8.0, as in - we are not fixing some complete showstoppers here.

Comment (VSadov) [reply src/coreclr/vm/jitinterface.cpp:6051]: I have created a "reduced" version of this. I think that is what we can consider for 8.0 - https://github.com/dotnet/runtime/pull/90234

I have removed all the JIT changes, but kept the added codegen test.

Comment (jkotas) [reply src/coreclr/vm/jitinterface.cpp:6051]: Sounds good.

**Comment by VSadov** [comment]
A reduced version of this affecting only run time behavior has been merged. 
For the further improvements for the JIT API in this area a tracking bug has been added - https://github.com/dotnet/runtime/issues/91016
